src-opam/opam.ml

```diff
@@ -47,6 +53,38 @@ let install_opam_from_source ?(add_default_link = true) ?(prefix = "/usr/local")
     prefix branch prefix branch prefix branch branch
   @@ maybe_link_opam add_default_link prefix branch
 
+(* Build opam in a separate worktree from an already cloned opam *)
+let install_opam_from_source_windows ?cyg ?(add_default_link = true)
+    ?(prefix = "/usr/local") ?(enable_0install_solver = false) ~branch ~hash ()
+    =
+  let cold_compiler = true and lib_pkg = true in
+  Windows.Cygwin.run_sh ?cyg
+    "cd /tmp/opam-sources && cp -P -R -p . ../opam-build-%s && cd \
+     ../opam-build-%s && git checkout %s && git config --global --add \
+     safe.directory /tmp/opam-build-%s"
+    branch branch hash branch
+  @@ (if cold_compiler then
+      Windows.Cygwin.run_sh ?cyg
+        {|cd /tmp/opam-build-%s && make cold CONFIGURE_ARGS="--enable-cold-check%s"|}
+        branch
+        (if enable_0install_solver then " --with-0install-solver" else "")
+     else
+       Windows.Cygwin.run_sh ?cyg "cd /tmp/opam-build-%s && make compiler"
+         branch
+       @@ (if lib_pkg then
+           Windows.Cygwin.run_sh ?cyg "cd /tmp/opam-build-%s && make lib-pkg"
+             branch
+          else empty)
+       @@ Windows.Cygwin.run_sh ?cyg
+            "cd /tmp/opam-build-%s && ./configure --enable-cold-check%s && make"
+            branch
+            (if enable_0install_solver then " --with-0install-solver" else ""))
+  @@ Windows.Cygwin.run_sh ?cyg
+       "cd /tmp/opam-build-%s && mkdir -p %s/bin && cp /tmp/opam-build-%s/opam \
+        %s/bin/opam-%s && chmod a+x %s/bin/opam-%s"
```

Comment:
I tried removing the build directory, but `rm` complained about an non-empty dir. Presumably, some file inside of it couldn't be removed (permission issue?). That's not too bad since I'm using a separate build stage.